[lldb/API] Mark SBValue with error as invalid

medismailben opened the pull request:
This patch fixes the return value of `SBValue::IsValid` if `GetError` returns a failing `SBError`.

That alignes better the expectation that an `SBValue` is invalid if it has an error.

adrian-prantl commented on lldb/test/API/python_api/sbvalue_is_valid/main.cpp:
```diff
@@ -0,0 +1,6 @@
+  int i = 42;
+  return 0; //% v1 = self.frame().EvaluateExpression("test"); v2 = self.frame().EvaluateExpression("i");
+  //% self.assertFalse(v1.IsValid())
+  //% self.assertTrue(v2.IsValid())
```
Once you have more than one line of comments, a regular (non inline) python tests is IMO more readable.

medismailben commented:
I fixed the formatting most of the failures. I only need to find the right way to return the errors through DAP: prior to this change, errors were returned as part of the value object string, since the object was still considered valid but now that it's marked invalid, the return string is empty. @JDevlieghere what would be the right approach here ?

medismailben committed:
[lldb/API] Mark SBValue with error as invalid

This patch fixes the return value of `SBValue::IsValid` if `GetError`
returns a failing `SBError`.

That alignes better the expectation that an `SBValue` is invalid if it has
an error.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

JDevlieghere commented:
I was under the impression that these two (`IsValid` and `GetError`) were intentionally decoupled, although I don't remember the reasoning. I vaguely remember this came up when @PortalPete was working on this. 

@adrian-prantl / @jimingham Do I misremember?

jimingham reviewed (changes requested):
This doesn't sound like a good idea to me.  IsValid means "will I get meaningful information back if I ask a question of this object".  You certainly want to know whether the Error state of the SBValue is in fact valid.  So an SBValue with valid error information should always be in the IsValid = true state.

jimingham commented:
Note, another way to do this would be to have default constructed SBValues have a Failed error state of "Nothing in me now".  Then you could dispense with SBValie.IsValid altogether, since there aren't any SBValues that don't at least have valid error states.

medismailben commented:
This behavior is counterintuitive. Once you have an `SBValue`, there isn’t really a meaningful "question" the user can ask of it besides calling `GetError`. Checking whether the underlying pointer is non-null is just an implementation detail.

In C++, `std::make_shared` can only fail in two cases: out-of-memory (`std::bad_alloc`) or if the constructor of the managed object throws. But since LLVM/LLDB are built with exceptions disabled, either case would immediately terminate the LLDB. So if LLDB is still running, the pointer you got is guaranteed to be valid.

That makes IsValid() in its current form effectively useless:  it never provides actionable information. Since we cannot remove `IsValid` from the API, I think we should make the it more natural and practical to use, without burdening user with redundant validity checks. That’s the reasoning behind my patch.
